### PR TITLE
Adding touchscreen functionality for the FT63X6 driver

### PR DESCRIPTION
**Pull request by gpambrozio**
# What does this implement/fix?

Touchscreen functionality for the FT63X6 driver (https://focuslcds.com/content/FT6236.pdf)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2352

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
# Might need to adjust display, this works well for the WT32-SC01 display: https://www.seeedstudio.com/ESP32-Development-board-WT32-SC01-p-4735.html
esphome:

esp32:
  board: m5stack-core2
  framework:
    type: arduino

external_components:
  - source: github://gpambrozio/esphome@nvds-touchscreen-optdisp-FT6336U
    components: [ ft63x6 ]

i2c:
  sda: 18
  scl: 19
  scan: false

globals:
  - id: calibration_x
    type: int
    restore_value: no
    initial_value: '0'
  - id: calibration_y
    type: int
    restore_value: no
    initial_value: '0'

font:
  - file: "gfonts://Roboto"
    id: roboto
    size: 20

color:
  - id: my_blue
    blue: 100%

output:
  - platform: ledc
    pin: GPIO23
    id: screen_led

light:
  - platform: monochromatic
    output: screen_led
    id: backlight
    default_transition_length: 0.5s
    name: 'Backlight'
    restore_mode: ALWAYS_ON

display:
  - platform: ili9xxx
    model: ST7796
    cs_pin: GPIO15
    dc_pin: GPIO21
    reset_pin: GPIO22
    rotation: 90
    pages:
      - id: calibration_page
        lambda: |-
          it.line(id(calibration_x) - 10, id(calibration_y), id(calibration_x) + 10, id(calibration_y), id(my_blue));
          it.line(id(calibration_x), id(calibration_y) - 10, id(calibration_x), id(calibration_y) + 10, id(my_blue));
          it.printf(it.get_width() / 2, it.get_height() / 2, id(roboto), TextAlign::CENTER, "%d x %d", id(calibration_x), id(calibration_y));

touchscreen:
  - platform: ft63x6
    on_touch:
      - logger.log:
          format: Touch at (%d, %d)
          args: [touch.x, touch.y]
      - globals.set:
          id: calibration_x
          value: !lambda 'return touch.x;'
      - globals.set:
          id: calibration_y
          value: !lambda 'return touch.y;'

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by gpambrozio** [comment]
@nielsnl68 I managed to fix my display and tested and it seems to work fine for me. I created this pr against your branch to make it easier to review. Would be probably good to either get this merged to your branch or alternatively I can also just create a new PR on esphome once your merges.

**Comment by nielsnl68** [comment]
Thanks @gpambrozio, between the old and my way of handling is where the touch control is executed.
In the past every touchscreen driver would need handle the touch events it self. 
In my proposal PR this last is now fully controlled by the Touchscreen class. The only thing the touchscreen driver like yours need to handle the virtual `touch_handle()` method and where you need to fill the `touches` list with the raw coordinators and finger ID of the touched point. Everything else will be handled in the touchscreen class.

In the `setup()`  you need to set the `this->x_raw_max_` and `this->y_raw_max_`  re placing the 
```
  this->x_resolution_ = 320;
  this->y_resolution_ = 480;
```

i hope you could make the needed changes. for a good example you can check the changes i made in this driver:
https://github.com/esphome/esphome/pull/4596/files#diff-fcd891cf0e47e772b11c3a78f4314ba381919328995a33e1a2cbe50ba42f619d

File: esphome/components/ft63x6/ft63x6.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  uint8_t touch_count;
+  TouchPointType tp[2];
+};
+
+struct FT63X6TouchscreenStore {
+  volatile bool touch;
+  ISRInternalGPIOPin pin;
+
+  static void gpio_intr(FT63X6TouchscreenStore *store);
+};
```
Comment (nielsnl68) [inline esphome/components/ft63x6/ft63x6.h:40]:
```suggestion
```

File: esphome/components/ft63x6/ft63x6.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+using namespace touchscreen;
+
+class FT63X6Touchscreen : public Touchscreen, public PollingComponent, public i2c::I2CDevice {
```
Comment (nielsnl68) [inline esphome/components/ft63x6/ft63x6.h:44]:
```suggestion
class FT63X6Touchscreen : public Touchscreen, public i2c::I2CDevice {
```

File: esphome/components/ft63x6/ft63x6.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+ public:
+  void setup() override;
+  void loop() override;
+  void update() override;
```
Comment (nielsnl68) [inline esphome/components/ft63x6/ft63x6.h:48]:
```suggestion
```

File: esphome/components/ft63x6/ft63x6.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  uint16_t x_resolution_;
+  uint16_t y_resolution_;
+
+  uint8_t read_touch_count_();
+  uint16_t read_touch_coordinate_(uint8_t coordinate);
+  uint8_t read_touch_id_(uint8_t id_address);
+
+  FT63X6TouchscreenStore store_;
+  FT63X6TouchPoint touch_point_;
+  bool touched_{false};
```
Comment (nielsnl68) [inline esphome/components/ft63x6/ft63x6.h:70]:
```suggestion
  void update_touches() override;

  InternalGPIOPin *interrupt_pin_{nullptr};
  GPIOPin *reset_pin_{nullptr};

    uint8_t read_touch_count_();
  uint16_t read_touch_coordinate_(uint8_t coordinate);
  uint8_t read_touch_id_(uint8_t id_address);
```

File: esphome/components/ft63x6/touchscreen.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        {
+            cv.GenerateID(): cv.declare_id(FT63X6Touchscreen),
+            cv.Optional(CONF_INTERRUPT_PIN): cv.All(
+                pins.internal_gpio_input_pin_schema
+            ),
+            cv.Optional(CONF_RESET_PIN): pins.gpio_output_pin_schema,
+        }
+    )
+    .extend(i2c.i2c_device_schema(0x38))
+    .extend(cv.COMPONENT_SCHEMA)
```
Comment (nielsnl68) [inline esphome/components/ft63x6/touchscreen.py:33]:
```suggestion
```

File: esphome/components/ft63x6/touchscreen.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+)
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    await i2c.register_i2c_device(var, config)
+    await touchscreen.register_touchscreen(var, config)
```
Comment (nielsnl68) [inline esphome/components/ft63x6/touchscreen.py:41]:
```suggestion
    await touchscreen.register_touchscreen(var, config)
    await i2c.register_i2c_device(var, config)
```

**Review by nielsnl68 — CHANGES_REQUESTED** [review]

File: esphome/components/ft63x6/ft63x6.h [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+ protected:
+  void hard_reset_();
+  uint8_t read_byte_(uint8_t addr);
+  void update_touches() override;
+
+  InternalGPIOPin *interrupt_pin_{nullptr};
+  GPIOPin *reset_pin_{nullptr};
+
+    uint8_t read_touch_count_();
```
Comment (nielsnl68) [inline esphome/components/ft63x6/ft63x6.h:36]:
```suggestion
  uint8_t read_touch_count_();
```

File: esphome/components/ft63x6/ft63x6.cpp [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+void FT63X6Touchscreen::update_touches() {
+  touch_point_.touch_count = read_touch_count_();
+
+  if (touch_point_.touch_count == 0) {
+    return;
+  }
+
+  touched_ = true;
+
```
Comment (nielsnl68) [inline esphome/components/ft63x6/ft63x6.cpp:57]:
```suggestion
```

**Review by nielsnl68 — CHANGES_REQUESTED** [review]

**Review by nielsnl68 — APPROVED** [review]

File: esphome/components/ft63x6/touchscreen.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    cv.Schema(
+        {
+            cv.GenerateID(): cv.declare_id(FT63X6Touchscreen),
+            cv.Optional(CONF_INTERRUPT_PIN): cv.All(
+                pins.internal_gpio_input_pin_schema
+            ),
+            cv.Optional(CONF_RESET_PIN): pins.gpio_output_pin_schema,
+        }
+    )
+    .extend(i2c.i2c_device_schema(0x38))
```
Comment (nielsnl68) [inline esphome/components/ft63x6/touchscreen.py:32]:
```suggestion
    ).extend(i2c.i2c_device_schema(0x38))
```

File: esphome/components/ft63x6/ft63x6.cpp [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+  if (this->reset_pin_ != nullptr) {
+    this->reset_pin_->setup();
+  }
+
+  this->hard_reset_();
+
+  // Get touch resolution
+  this->x_raw_max_= 320;
+  this->y_raw_max_= 480;
```
Comment (nielsnl68) [inline esphome/components/ft63x6/ft63x6.cpp:45]:
```suggestion
  this->x_raw_max_ = 320;
  this->y_raw_max_ = 480;
```

**Review by nielsnl68 — CHANGES_REQUESTED** [review]

File: esphome/components/ft63x6/ft63x6.cpp [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    this->reset_pin_->setup();
+  }
+
+  this->hard_reset_();
+
+  // Get touch resolution
+  this->x_raw_max_ = 320;
+  this->y_raw_max_ = 480;
+}
+
```
Comment (nielsnl68) [inline esphome/components/ft63x6/ft63x6.cpp:46]:
```suggestion
```

**Review by nielsnl68 — CHANGES_REQUESTED** [review]

**Pull request merged**